pkg/ccl/changefeedccl/kvevent/chunked_event_queue.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Licensed as a CockroachDB Enterprise file under the Cockroach Community
+// License (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package kvevent
+
+import "sync"
+
+const bufferEventChunkArrSize = 128
+
+// bufferEventChunkQueue is a queue implemented as a linked-list of bufferEntry.
+type bufferEventChunkQueue struct {
+	head, tail *bufferEventChunk
+}
+
+func (l *bufferEventChunkQueue) enqueue(e Event) {
+	if l.tail == nil {
+		chunk := newBufferEntryChunk()
+		l.head, l.tail = chunk, chunk
+		l.head.push(e) // guaranteed to insert into new chunk
+		return
+	}
+
+	if !l.tail.push(e) {
+		chunk := newBufferEntryChunk()
+		l.tail.next = chunk
+		l.tail = chunk
+		l.tail.push(e) // guaranteed to insert into new chunk
+	}
+}
+
+func (l *bufferEventChunkQueue) empty() bool {
+	return l.head == nil || l.head.empty()
+}
+
+func (l *bufferEventChunkQueue) dequeue() (e Event, ok bool) {
+	if l.head == nil {
+		return Event{}, false
+	}
+
+	e, ok, consumedAllFromChunk := l.head.pop()
+
+	if consumedAllFromChunk {
```

Comment:
It would be nice if we didn't have to do these recursive calls.
I think the code (and datastructures) could be simplified.
Currently, we have:
```
type bufferEventChunkQueue struct {
	head, tail *bufferEventChunk
}
type bufferEventChunk struct {
	events     [bufferEventChunkArrSize]Event
	head, tail int32
	next       *bufferEventChunk // linked-list element
}
```

It's a bit sad that each chunk has to carry head/tail counters.
I think those counters can be lifted to bufferEventQueue
(drop the head/tail from bufferEventChunk):
```
type bufferEventChunkQueue struct {
  head, tail *bufferEventChunk
  headChunkIdx, tailChunkIdx int      
}
```

Enqueue becomes pretty easy:
```
func (l *bufferEventChunkQueue) enqueue(e Event) {
   if l.tail == nil || l.tailChunkIdx == bufferEventChunkArrSize {
      // Add new chunk; this either creates or updates l.tail
   }
   l.tail.events[l.tailChunkIdx] = e
   l.tailChunkIdx++
}
```

Dequeue is similarly easy
```
func (l *bufferEventChunkQueue) dequeue() (e Event, ok bool) {
   // You always dequeue from head
   e = l.head.events[l.headChunkIdx]
  l.headChunkIdx++
  if l.headChunkIdx == bufferEventChunkArrSize {
     // free l.head
  }
}
```